File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int memo(int n,vector<int>&dp)
    {
        if (n == 0 || n == 1)
            return 1;
        if(dp[n]!=-1)return dp[n];
        int l = memo(n - 1,dp);
        int r = memo(n - 2,dp);
        // cout << "n :" << n << " l: " << l << " r: " << r << endl;
        return dp[n] = l + r;
    }
    int climbStairs(int n) {
        vector<int>dp(n+1,-1);
        return memo(n,dp);
    }
};

File: 0070-climbing-stairs/0070-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        int []dp = new int[n+1];
        Arrays.fill(dp,-1);
        return climb(n,dp);
    }
    public int climb(int n,int dp[]){
        if(n==1 || n==0)
            return 1;
        if(dp[n]!=-1)return dp[n];
        int one = climb(n-1,dp);
        int two = climb(n-2,dp);
        dp[n] = one+two;
        return dp[n];
    }
    // public int climbStairs(int n) {
    //     int []ans = {0};
    //     int step = 0;
    //     climb(n,step,ans);
    //     return ans[0];
    // }
    // public void climb(int n,int step,int []ans){
    //     if(step >= n){
    //         if(step==n)ans[0]++;
    //         return;
    //     }
    //     climb(n,step+1,ans);
    //     if((n-step)>=2)
    //     climb(n,step+2,ans);
    // }
}